gobblin-utility/src/main/java/gobblin/util/FileListUtils.java

```diff
@@ -57,15 +57,26 @@ public boolean accept(Path path) {
    */
   public static List<FileStatus> listFilesRecursively(FileSystem fs, Path path, PathFilter fileFilter)
       throws IOException {
-    return listFilesRecursivelyHelper(fs, Lists.<FileStatus> newArrayList(), fs.getFileStatus(path), fileFilter);
+    return listFilesRecursivelyHelper(fs, Lists.<FileStatus> newArrayList(), fs.getFileStatus(path), fileFilter, false);
+  }
+
+  /**
+   * Helper method to list out all files under a specified path. The specified {@link PathFilter} is treated as a file
+   * filter, that is it is only applied to file {@link Path}s.
+   */
+  public static List<FileStatus> listFilesRecursively(FileSystem fs, Path path, PathFilter fileFilter,
+      boolean applyFilterToDirectories) throws IOException {
+    return listFilesRecursivelyHelper(fs, Lists.<FileStatus> newArrayList(), fs.getFileStatus(path), fileFilter,
+        applyFilterToDirectories);
   }
 
   @SuppressWarnings("deprecation")
   private static List<FileStatus> listFilesRecursivelyHelper(FileSystem fs, List<FileStatus> files,
-      FileStatus fileStatus, PathFilter fileFilter) throws FileNotFoundException, IOException {
+      FileStatus fileStatus, PathFilter fileFilter, boolean applyFilterToDirectories)
+      throws FileNotFoundException, IOException {
     if (fileStatus.isDir()) {
-      for (FileStatus status : fs.listStatus(fileStatus.getPath())) {
-        listFilesRecursivelyHelper(fs, files, status, fileFilter);
+      for (FileStatus status : fs.listStatus(fileStatus.getPath(), applyFilterToDirectories ? fileFilter : NO_OP_PATH_FILTER)) {
+        listFilesRecursivelyHelper(fs, files, status, fileFilter, applyFilterToDirectories);
```

Comment:
if `status` is a file you can directly add it to `files` here. Otherwise it will evaluate the filter on the file twice.